test/e2e/cluster_deletion.go

```diff
@@ -0,0 +1,357 @@
+			PostMachinesProvisioned: func() {
+				if input.PostMachinesProvisioned != nil {
+					input.PostMachinesProvisioned(input.BootstrapClusterProxy, namespace.Name, clusterName)
+				}
+			},
+		}, clusterResources)
+
+		relevantKinds := sets.Set[string]{}
+
+		for _, phaseInput := range input.ClusterDeletionPhases {
+			relevantKinds = relevantKinds.
+				Union(phaseInput.KindsWithFinalizer).
+				Union(phaseInput.Kinds)
+		}
+
+		// Get all objects relevant to the test by filtering for the kinds of the phases.
+		graph, err := clusterctlcluster.GetOwnerGraph(ctx, namespace.GetName(), input.BootstrapClusterProxy.GetKubeconfigPath(), func(u unstructured.Unstructured) bool {
+			return relevantKinds.Has(u.GetKind())
+		})
+		Expect(err).ToNot(HaveOccurred())
+
+		for i, phase := range input.ClusterDeletionPhases {
+			// Iterate over the graph and find all objects relevant for the phase.
+			// * Objects matching the kind except for Machines (see below)
+			// * Machines which are owned by one of the kinds which are part of the phase
+			allKinds := phase.KindsWithFinalizer.Union(phase.Kinds)
+			allKindsWithoutMachine := allKinds.Clone().Delete("Machine")
+
+			for _, node := range graph {
+				// Only add objects which match a defined kind for the phase.
+				if !allKinds.Has(node.Object.Kind) {
+					continue
+				}
+
+				// Special handling for machines: only add machines which are owned by one of the given kinds.
+				if node.Object.Kind == "Machine" {
+					isOwnedByKind := false
+					for _, owner := range node.Owners {
+						if allKindsWithoutMachine.Has(owner.Kind) {
+							isOwnedByKind = true
+							break
+						}
+					}
+					// Ignore Machines which are not owned by one of the given kinds.
+					if !isOwnedByKind {
+						continue
+					}
+				}
+
+				obj := new(unstructured.Unstructured)
+				obj.SetAPIVersion(node.Object.APIVersion)
+				obj.SetKind(node.Object.Kind)
+				obj.SetName(node.Object.Name)
+				obj.SetNamespace(namespace.GetName())
+
+				phase.objects = append(phase.objects, obj)
+
+				// If the object's kind is part of kindsWithFinalizer, then additionally the object to the finalizedObjs slice.
+				if phase.KindsWithFinalizer.Has(obj.GetKind()) {
+					objectsWithFinalizer = append(objectsWithFinalizer, obj)
+				}
+			}
+
+			// Update the phase in input.ClusterDeletionPhases
+			input.ClusterDeletionPhases[i] = phase
+		}
+
+		// Update all objects in objectsWithFinalizers and add the finalizer for this test to control when these objects vanish.
+		addFinalizer(ctx, input.BootstrapClusterProxy.GetClient(), finalizer, objectsWithFinalizer...)
+
+		// Trigger the deletion of the Cluster.
+		framework.DeleteCluster(ctx, framework.DeleteClusterInput{
+			Cluster: clusterResources.Cluster,
+			Deleter: input.BootstrapClusterProxy.GetClient(),
+		})
+
+		var objectsDeleted, objectsInDeletion []client.Object
+
+		for i, phase := range input.ClusterDeletionPhases {
```

Comment:
Let's add a log line for each phase, so we can figure out what is going on in case of errors